dbcon/joblist/jlf_tuplejoblist.cpp

```diff
@@ -3356,6 +3932,10 @@ void joinTablesInOrder(uint32_t largest, JobStepVector& joinSteps, TableInfoMap&
         largeIndices.push_back(getKeyIndex(*k2, largeSideRG));
       }
 
+      // Try to restore `circular join edge` if possible.
+      tryToRestoreJoinEdges(jobInfo, info, largeSideRG, smallIndices, largeIndices, traces, joinIdIndexMap,
+                            smallSideIndex);
+      typeless.push_back(info->fJoinData.fTypeless);
```

Comment:
This is to allow the join to become typeless if we restored an edge that triggers typeless join?

Reply:
Right, to have a multiple keys join, the join type should be a `typeless`. This flag is set  in function `tryToRestoreJoinEdges` if we match keys in the smallside/largeside rowgroups.
```
// Mark as tupeless for multiple keys join.
 joinInfo->fJoinData.fTypeless = true;
```